### PR TITLE
Avoid race condition between transaction commit and cache clear

### PR DESCRIPTION
**Pull request by PeterJCLaw**
(inline comment duplicated here)

Only clear the cache when the current transaction commits.

While clearing the cache earlier than that is valid, it is insufficient to ensure cache consistency. There is a possible race between two transactions as follows:

1. Transaction 1: modifies model (and thus clears cache)
2. Transaction 2: queries cache, which misses, so it populates the cache from the database, picking up the unmodified model
3. Transaction 1: commits, without further signal to the cache

At this point the cache contains the _original_ value of the model, which is out of step with the database.

To avoid this we delay clearing the cache until the transaction commits. While this does leave a small window after the transaction has committed but before the cache has cleared, that is better than leaving the cache incorrect until the model is next updated.

### CODE REVIEW

**Comment by lamby** [comment]
> At this point the cache contains the original value of the model, which is out of step with the database. 

<3

File: cache_toolbox/core.py [hunk 1]
```diff
@@ -96,7 +96,26 @@ def delete_instance(model, *instance_or_pk):
...
+    #   Transaction 1: commits, without further signal to the cache
+    #
+    # At this point the cache contains the _original_ value of the model, which
+    # is out of step with the database.
+    # To avoid this we delay clearing the cache until the transaction commits.
+    # While this does leave a small window after the transaction has committed
+    # but before the cache has cleared, that is better than leaving the cache
+    # incorrect until the model is next updated.
+
+    transaction.on_commit(lambda: cache.delete_many(
```
Comment (lamby) [inline cache_toolbox/core.py:116]: Can we pass the method and args without using a `lambda`?

**Review by lamby — CHANGES_REQUESTED** [review]

File: cache_toolbox/core.py (hunk 1 above)
Comment (PeterJCLaw) [reply cache_toolbox/core.py:116]: I don't think so -- [the docs](https://docs.djangoproject.com/en/1.11/topics/db/transactions/#performing-actions-after-commit) suggest using a `lambda` and it looks like the signature of `on_commit` only allows a `func` (it also takes a database to look at, which I suspect is why it doesn't also take function arguments).

Comment (lamby) [reply cache_toolbox/core.py:116]: Ah, I think I was confusing it with eventlet/gunicorn's "call me ~~maybe~~ later" calling convention. Thanks!

**Comment by lamby** [comment]
Released as `0.3.1`. Thanks!

**Pull request merged**